backend/src/test/java/com/google/backend/servlets/LoginServletTest.java

```diff
@@ -0,0 +1,131 @@
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.Mockito.*;
+import org.mockito.MockitoAnnotations;
+
+/** Contain tests for methods in {@link LoginServlet} class. */
+@RunWith(JUnit4.class)
+public class LoginServletTest {
+  @Mock HttpServletRequest request;
+  @Mock HttpServletResponse response;
+
+  private static final LoginServlet loginServlet = new LoginServlet();
+  StringWriter stringWriter = new StringWriter();
+  private final LocalServiceTestHelper helper = new LocalServiceTestHelper(
+    new LocalDatastoreServiceTestConfig(), new LocalUserServiceTestConfig());
+
+  private static final String AUTH_DOMAIN_NAME = "localhost";
+  private static final String CONTENT_TYPE = "text/html;";
+  private static final String PROPERTY_NAME = "email";
+  private static final String TEST_EMAIL = "test@example.com";
+  private static final String REDIRECT_URL = "/";
+  private static final String LOGIN_STATUS = "logged in";
+  private static final String LOGOUT_STATUS = "stranger";
+  private static final String ENTITY_KIND = "User";
+  private static final int MAX_QUERIES = 1000;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    when(response.getWriter()).thenReturn(new PrintWriter(stringWriter));
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void doGet_GetsCorrectResponseLoggedIn() throws IOException, ServletException {
+    helper.setEnvIsLoggedIn(true).setEnvEmail(TEST_EMAIL).setEnvAuthDomain(AUTH_DOMAIN_NAME);
+    UserService userService = UserServiceFactory.getUserService();
+    String logoutUrl = userService.createLogoutURL(REDIRECT_URL);
+    String loginResponse = LOGIN_STATUS + "\n" + logoutUrl;
+
+    loginServlet.doGet(request, response);
+
+    verify(response).setContentType(CONTENT_TYPE);
+    assertEquals(stringWriter.getBuffer().toString().trim(), loginResponse);
+  }
+
+  @Test
+  public void doGet_UserDoesNotExistCreateNewEntity() throws IOException, ServletException {
+    helper.setEnvIsLoggedIn(true).setEnvEmail(TEST_EMAIL).setEnvAuthDomain(AUTH_DOMAIN_NAME);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    loginServlet.doGet(request, response);
+    
+    Query query =
+          new Query(ENTITY_KIND).setFilter(
+            new Query.FilterPredicate(PROPERTY_NAME, Query.FilterOperator.EQUAL, TEST_EMAIL));
+    Entity result = datastore.prepare(query).asSingleEntity();
+    assertNotNull(result);
+  }
+
+  @Test
+  public void doGet_UserExistsNoNewEntity() throws IOException, ServletException {
```

Comment:
nit: split more clearly into arrange, act, assert (like 3 blocks?)

Reply:
i think i added an extra line in there haha